### PR TITLE
refactor(grpc): add a flag to omit call to grpc metadata for client-originated error

### PR DESCRIPTION
**Pull request by alevenberg**
https://github.com/googleapis/google-cloud-cpp/issues/13724

Add an error_origin param to GetRequestMetadataFromContext. Set the default parameter as `kErrorOrigin::kUnknown` since that is the current behavior.

This flag will be set by the ErrorOrigin(Status s) call.

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14119)

### CODE REVIEW

File: google/cloud/internal/status_utils.h [hunk 1]
```diff
@@ -23,7 +23,9 @@ namespace cloud {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
-bool IsClientOrigin(Status const& status);
+enum class ErrorOrigin { kUnknown, kClient };
+
+ErrorOrigin IsClient(Status const& status);
```
Comment (devbww) [inline google/cloud/internal/status_utils.h:28]: `IsBlah()` still sounds like something to be used in a boolean context.

I would suggest reusing the same name.  That is,
```
ErrorOrigin ErrorOrigin(Status const&);
```
I believe you will find that is possible.  So, `if (ErrorOrigin(status) != ErrorOrigin::kClient) ...`.

[BTW, `IsClient()` doesn't appear to be used anywhere (except the test).  I presume that comes later?]

File: google/cloud/internal/grpc_request_metadata.cc [hunk 2]
```diff
@@ -66,10 +68,15 @@ void AddContextMetadata(grpc::ClientContext const& context,
 
 }  // namespace
 
-RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context) {
+RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context,
+                                          internal::ErrorOrigin error_origin) {
   RpcMetadata metadata;
   AddContextMetadata(context, metadata);
-  AddServerRequestMetadata(context, metadata);
+  // If there is a client originating error, the request never made it to the
```
Comment (devbww) [inline google/cloud/internal/grpc_request_metadata.cc:75]: Perhaps `s/is a client originating error/was a client-originated error/`.

**Review by devbww — COMMENTED** [review]
PS: The PR description (and perhaps title) looks like it could be with an update.

File: google/cloud/internal/grpc_request_metadata.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/internal/grpc_request_metadata.cc:75]: Done.

File: google/cloud/internal/status_utils.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/internal/status_utils.h:28]: Done. And yes, it will be used later.

Comment (alevenberg) [reply google/cloud/internal/status_utils.h:28]: I was getting 

```
google/cloud/internal/status_utils.cc:23:1: error: 'ErrorOrigin' does not name a type
   23 | ErrorOrigin ErrorOrigin(Status const& status) {
      | ^~~~~~~~~~~
```

So renamed to GetErrorOrigin. Lmk what you think

File: google/cloud/internal/grpc_request_metadata.cc [hunk 3]
```diff
@@ -66,10 +68,15 @@ void AddContextMetadata(grpc::ClientContext const& context,
...
 }  // namespace
 
-RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context) {
+RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context,
+                                          internal::ErrorOrigin error_origin) {
   RpcMetadata metadata;
   AddContextMetadata(context, metadata);
-  AddServerRequestMetadata(context, metadata);
+  // If there was a client-originated error, the request never made it to the
+  // server metadata. Therefore, it doesn't exist so we should skip the call.
```
Comment (devbww) [inline google/cloud/internal/grpc_request_metadata.cc:76]: Perhaps clarify "it".

File: google/cloud/internal/status_utils.h (hunk 1 above)
Comment (devbww) [reply google/cloud/internal/status_utils.h]: Sounds good.  Sorry for being misleading.

**Review by devbww — APPROVED** [review]
The PR description still refers to #14104, which is closed.

File: google/cloud/internal/grpc_request_metadata.cc (hunk 3 above)
Comment (dbolduc) [reply google/cloud/internal/grpc_request_metadata.cc]: > "the request never made it to the server metadata"

I do not understand what this means. I think you are trying to say "the initial server metadata was never received by the client".

File: google/cloud/internal/grpc_request_metadata.h [hunk 4]
```diff
@@ -27,7 +28,8 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
 /// Return interesting bits of metadata stored in the client context.
-RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context);
+RpcMetadata GetRequestMetadataFromContext(grpc::ClientContext const& context,
+                                          internal::ErrorOrigin error_origin);
```
[the remark this answers is not in the document]
Comment (dbolduc) [reply google/cloud/internal/grpc_request_metadata.h:32]: What was wrong with `GprcMetadataView` ?

File: google/cloud/internal/grpc_request_metadata.cc (hunk 3 above)
Comment (alevenberg) [reply google/cloud/internal/grpc_request_metadata.cc:76]: Done

File: google/cloud/internal/status_utils.h (hunk 1 above)
Comment (alevenberg) [reply google/cloud/internal/status_utils.h]: All good. I think the point was valid either way. Thanks!

File: google/cloud/internal/status_utils.cc [hunk 5]
```diff
@@ -20,11 +20,13 @@ namespace cloud {
 GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 namespace internal {
 
-bool IsClientOrigin(Status const& status) {
+ErrorOrigin GetErrorOrigin(Status const& status) {
```
Comment (dbolduc) [inline google/cloud/internal/status_utils.cc:23]: comment: I think returning a `bool` was fine for readability.

```cc
if (IsClientOrigin(status)) {
  GetRequestMetadata(context, kWithoutHeaders);
}
```

Returning a `bool` is a poor choice when the function name doesn't give you any hints. e.g.: https://github.com/googleapis/google-cloud-cpp/blob/a77950e488c3be12832aa5efece7a0889ff47a08/google/cloud/retry_policy.h#L61

... where I have to look every time whether `true` means retry, or `false` means retry.

Comment (dbolduc) [reply google/cloud/internal/status_utils.cc:23]: I see that you are trying to unify two separate things, but I don't think it makes sense. If we need two enums, one for `ErrorOrigin` and one for `GrpcMetadataView`, so be it.

**Review by dbolduc — CHANGES_REQUESTED** [review]
I think the `GetRequestMetadata` API needs to be fixed.

File: google/cloud/internal/grpc_request_metadata.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/internal/grpc_request_metadata.h]: Discussed offline and addressed all open comments.

File: google/cloud/internal/async_read_write_stream_impl.h [hunk 6]
```diff
@@ -19,8 +19,8 @@
 #include "google/cloud/completion_queue.h"
 #include "google/cloud/grpc_error_delegate.h"
 #include "google/cloud/internal/completion_queue_impl.h"
+#include "google/cloud/internal/grpc_metadata_view.h" #include "google/cloud/options.h"
```
Comment (dbolduc) [inline google/cloud/internal/async_read_write_stream_impl.h:22]: ?

File: google/cloud/internal/async_streaming_read_rpc_impl.h [hunk 7]
```diff
@@ -121,7 +122,7 @@ class AsyncStreamingReadRpcImpl : public AsyncStreamingReadRpc<Response> {
   }
 
   RpcMetadata GetRequestMetadata() const override {
-    return GetRequestMetadataFromContext(*context_);
+    return GetRequestMetadataFromContext(*context_, ErrorOrigin::kUnknown);
```
Comment (dbolduc) [inline google/cloud/internal/async_streaming_read_rpc_impl.h]: ?

File: google/cloud/internal/grpc_request_metadata_test.cc [hunk 8]
```diff
@@ -46,6 +49,24 @@ TEST(GrpcRequestMetadata, GetRequestMetadataFromContext) {
                                                 Pair("trailer2", "value4")));
 }
 
+TEST(GrpcRequestMetadata,
+     GetRequestMetadataFromContextWhenInitialMetadataIsNotReady) {
+  auto const server_metadata =
+      RpcMetadata{{{"header1", "value1"}, {"header2", "value2"}},
+                  {{"trailer1", "value3"}, {"trailer2", "value4"}}};
```
Comment (dbolduc) [inline google/cloud/internal/grpc_request_metadata_test.cc:56]: nit: unused

File: google/cloud/internal/streaming_write_rpc_impl.h [hunk 9]
```diff
@@ -16,8 +16,8 @@
 #define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_STREAMING_WRITE_RPC_IMPL_H
 
 #include "google/cloud/grpc_error_delegate.h"
+#include "google/cloud/internal/grpc_metadata_view.h" #include "google/cloud/internal/streaming_write_rpc.h"
```
Comment (dbolduc) [inline google/cloud/internal/streaming_write_rpc_impl.h]: ?

**Pull request merged**